# PROJECT REPORT#
* This project used data from The United States National Climatic Data Center, This data is the weather in Dallas from 01/1895 to 01/2017.
* Table meaning:
    * PCP - Precipitation Index
    * TAVG - Temperature Index
    * TMIN - Minimum Temperature Index
    * TMAX - Maximum Temperature Index
    * PDSI - Palmer Drought Severity Index
    * PHDI - Palmer Hydrological Drought Index
    * ZNDX - Palmer Z-Index
    * PMDI - Modified Palmer Drought Severity Index
    * CDD - Cooling Degree Days
    * HDD - Heating Degree Days
    * SPnn - Standard Precipitation Index

### QUESTION 1: (Using Pandas, Numpy, matplotlib)
How the temp, PCP, HDD, and CDD have been changed from 1895-2017? 

In [2]:
import pandas as pd
import numpy as np
##import matplotlib.pyplot as plt
##import unicodecsv


df = pd.read_csv('TexasTemp.csv', index_col=0)

df.head()

Division  YearMonth  TAVG  TMIN  TMAX   PCP  PDSI  PHDI  ZNDX  \
StateCode                                                                  
41                0     189501  44.6  31.2  57.9  1.30  0.32  0.32  0.97   
41                0     189502  38.5  26.7  50.2  1.90  0.95  0.95  1.99   
41                0     189503  56.4  43.1  69.7  0.88  0.56  0.56 -0.88   
41                0     189504  64.8  50.0  79.6  1.38  0.26  0.26 -0.73   
41                0     189505  69.7  57.8  81.6  4.96  1.20  1.20  2.89   

           PMDI    CDD    HDD  SP01   SP02   SP03   SP06   SP09   SP12   SP24  
StateCode                                                                      
41         0.32   10.0  574.0 -0.13 -99.99 -99.99 -99.99 -99.99 -99.99 -99.99  
41         0.95    0.0  686.0  0.48   0.18 -99.99 -99.99 -99.99 -99.99 -99.99  
41         0.56   42.0  245.0 -0.99  -0.35  -0.36 -99.99 -99.99 -99.99 -99.99  
41        -0.51  108.0   55.0 -0.76  -1.18  -0.67 -99.99 -99.99 -99.99 -99.99  
41         1.20  211.0   10.0  1.19   0.47   0.08 -99.99 -99.99 -99.99 -99.99

In [4]:
df.tail()

Division  YearMonth  TAVG  TMIN  TMAX   PCP  PDSI  PHDI  ZNDX  \
StateCode                                                                  
41                0     201609  77.6  66.7  88.5  2.99  5.13  5.13 -0.48   
41                0     201610  70.9  57.7  84.1  0.69  3.63  3.63 -2.90   
41                0     201611  61.1  49.6  72.5  2.38  3.15  3.15 -0.31   
41                0     201612  49.6  38.9  60.3  2.14  2.87  2.87  0.14   
41                0     201701  50.8  38.7  62.9  2.32  2.70  2.70  0.35   

           PMDI    CDD    HDD  SP01  SP02  SP03  SP06  SP09  SP12  SP24  
StateCode                                                                
41         4.90  450.0    0.0  0.15  1.58  0.93  1.74  1.42  2.14  2.90  
41         2.28  258.0   13.0 -1.44 -0.80  0.55  0.68  1.12  1.32  2.84  
41         1.68   98.0  122.0  0.61 -0.56 -0.43  0.38  1.30  1.11  2.94  
41         1.45   19.0  387.0  0.55  0.63 -0.29  0.35  1.16  1.03  2.30  
41         1.39   39.0  369.0  0.95  0.89  0.88  0.90  0.95  1.31  2.30

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 41 to 41
Data columns (total 19 columns):
Division     1465 non-null int64
YearMonth    1465 non-null int64
TAVG         1465 non-null float64
TMIN         1465 non-null float64
TMAX         1465 non-null float64
PCP          1465 non-null float64
PDSI         1465 non-null float64
PHDI         1465 non-null float64
ZNDX         1465 non-null float64
PMDI         1465 non-null float64
CDD          1465 non-null float64
HDD          1465 non-null float64
SP01         1465 non-null float64
SP02         1465 non-null float64
SP03         1465 non-null float64
SP06         1465 non-null float64
SP09         1465 non-null float64
SP12         1465 non-null float64
SP24         1465 non-null float64
dtypes: float64(17), int64(2)
memory usage: 228.9 KB


This table below shows:
There is 1465 months  from 01/1895 to 01/2017.

The average temperature in 1465 months is 64.77 degree (Fahrenheit), the lowest average temperature is 36.6 degree, and the highest average temperature is 88.2 degree. Moreover, we can see that the lowest temperature from 1895-2017 is 24.1 degree and the highest temperature is 101.6 degree.

The average PHP in 1465 months is 2.26, the lowest is 0.02, and the highest is 9.05

In [6]:
df.describe()

Division      YearMonth         TAVG         TMIN         TMAX  \
count    1465.0    1465.000000  1465.000000  1465.000000  1465.000000   
mean        0.0  195560.694198    64.772355    52.330512    77.204505   
std         0.0    3525.377025    13.506872    13.648677    13.480776   
min         0.0  189501.000000    36.600000    24.100000    46.800000   
25%         0.0  192507.000000    52.000000    39.500000    64.700000   
50%         0.0  195601.000000    65.400000    52.700000    78.600000   
75%         0.0  198607.000000    77.900000    66.100000    89.800000   
max         0.0  201701.000000    88.200000    74.700000   101.600000   

               PCP         PDSI         PHDI         ZNDX         PMDI  \
count  1465.000000  1465.000000  1465.000000  1465.000000  1465.000000   
mean      2.265154    -0.142314    -0.065659    -0.025317    -0.083208   
std       1.276177     2.958140     3.093609     1.973357     2.955770   
min       0.020000    -8.050000    -8.050000    -5.570000    -8.050000   
25%       1.300000    -2.360000    -2.490000    -1.460000    -2.380000   
50%       2.070000    -0.340000     0.370000    -0.210000    -0.030000   
75%       2.950000     2.150000     2.400000     1.220000     2.160000   
max       9.050000     8.230000     8.230000     8.330000     8.230000   

               CDD          HDD         SP01         SP02         SP03  \
count  1465.000000  1465.000000  1465.000000  1465.000000  1465.000000   
mean    221.707850   169.031399     0.002526    -0.066280    -0.136437   
std     214.889381   203.507070     0.989223     2.792369     3.823301   
min       0.000000     0.000000    -3.040000   -99.990000   -99.990000   
25%      21.000000     0.000000    -0.700000    -0.690000    -0.700000   
50%     131.000000    57.000000     0.010000    -0.020000    -0.030000   
75%     428.000000   331.000000     0.690000     0.720000     0.710000   
max     750.000000   834.000000     3.090000     3.090000     3.090000   

              SP06         SP09         SP12         SP24  
count  1465.000000  1465.000000  1465.000000  1465.000000  
mean     -0.339823    -0.545672    -0.752348    -1.565973  
std       5.916969     7.437500     8.691142    12.476734  
min     -99.990000   -99.990000   -99.990000   -99.990000  
25%      -0.720000    -0.740000    -0.720000    -0.660000  
50%       0.000000     0.010000     0.010000     0.040000  
75%       0.690000     0.670000     0.680000     0.640000  
max       3.010000     2.880000     3.090000     3.080000

This code below changed the fomart of Yeahmothe from int to datetime format:

In [5]:
from datetime import datetime as dt
import unicodecsv

with open('TexasTemp.csv', 'rb') as f:
    reader = unicodecsv.DictReader(f)
    ymdates = list(reader)

def parse_date(date):
    if date == '':
        return None
    else:
        return dt.strptime(date, '%Y%m')
for ymdate in ymdates:
    ymdate['YearMonth'] = parse_date(ymdate['YearMonth'])
ymdates[0]

NameError: name 'unicodecsv' is not defined

This is a data table, and chart of average temperature, max temp, and min temp by each month in from 01/1895 to 01/2017:

In [ ]:
df[["YearMonth", "TAVG","TMIN","TMAX"]].groupby(['YearMonth'], as_index=False).mean().sort_values(by='YearMonth', ascending=True)

In [ ]:
import matplotlib.pyplot as plt

plt.plot(df["YearMonth"], df["TAVG"])
plt.xlabel('Time(YearMonth)')
plt.ylabel('Temp')
plt.title('TempvsTime')
plt.show()

This plot show the change of temperature in each month from 01/1895 to 01/2017. From this plot, we can see that the temp have been increased. From 19300s to 19400s, the temperature seem to droped to lowest temperatures in few months. However, in the summer of each years, we can see that the temperatures of those months had been increase litle bit. In this case, this is the effect of global warming from Pre-industrial era

In [ ]:
df[["YearMonth", "CDD", "HDD"]].groupby(['YearMonth'], as_index=False).mean().sort_values(by='YearMonth', ascending=True)

In [ ]:
import matplotlib.pyplot as plt

plt.plot(df["YearMonth"], df["CDD"])
plt.xlabel('Time(YearMonth)')
plt.ylabel('CDD')
plt.title('CDDvsTime')
plt.figure(figsize=(20,10))
plt.show()

In [ ]:
import matplotlib.pyplot as plt

plt.plot(df["YearMonth"], df["HDD"])
plt.xlabel('Time(YearMonth)')
plt.ylabel('HDD')
plt.title('HDDvsTime')
plt.figure(figsize=(20,10))
plt.show()

In [ ]:
df[["YearMonth", "PCP"]].groupby(['YearMonth'], as_index=False).mean().sort_values(by='YearMonth', ascending=True)

In [ ]:
import matplotlib.pyplot as plt

plt.plot(df["YearMonth"], df["PCP"])
plt.xlabel('Time(YearMonth)')
plt.ylabel('PCP')
plt.title('PCPvsTime')
plt.figure(figsize=(20,10))
plt.show()

### QUESTION 2 (Using MapReduce)
*What is the month has highest temp from 1895 - 2017?

    *In order to run TexasTemp Dataset on Mapreduce, I edit the TexasTemp.csv file by removing the first row "the name of collum" then save it to a new file "TexasTempNone.csv"

In [3]:
%%file maxTemp.py

from mrjob.job import MRJob
from mrjob.step import MRStep

class maxTemp(MRJob):
    def steps(self):
        return [
            MRStep(mapper=self.mapper_get_ratings,
                 reducer=self.reducer_count_ratings),
           MRStep(mapper=self.mapper_passthrough,
                 reducer = self.reducer_find_max)
        ]

    def mapper_get_ratings(self, _, line):
        (StateCode, Division, YearMonth, tavg, tmin, tmax, pcp, pdsi, phdi, zndx, pmdi,cdd, hdd, sp01, sp02, sp03, sp06, sp09, sp12, sp24) = line.split(",")
        yield YearMonth,float(tmax)
              
    def mapper_passthrough(self, key, values):
        yield key, values    
    
    def reducer_count_ratings(self, key, values):
        yield None, (max(values), key)
        
    def reducer_find_max(self, key, values):
        yield  max(values)
    
if __name__ == '__main__':
    maxTemp.run()

Overwriting maxTemp.py


In [4]:
!python maxTemp.py TexasTempNone.csv

No configs found; falling back on auto-configuration
Creating temp directory /var/folders/8h/_r5sr0qn5kb7bfj_9dmvj9hh0000gn/T/maxTemp.longdang2709.20170508.041352.894182
Running step 1 of 2...
Running step 2 of 2...
Streaming final output from /var/folders/8h/_r5sr0qn5kb7bfj_9dmvj9hh0000gn/T/maxTemp.longdang2709.20170508.041352.894182/output...
101.6	"201108"
Removing temp directory /var/folders/8h/_r5sr0qn5kb7bfj_9dmvj9hh0000gn/T/maxTemp.longdang2709.20170508.041352.894182...


The highest temperature is 101.6 in 08/2011

### QUESTION 3: (Using SprakPy)
What is the most common average temperature from 1895-2017?
In order to work with SprakPy, I activated the python 2.7

In [1]:
%%file commonTemp.py

from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("PopularTemp")
sc = SparkContext(conf = conf)

lines = sc.textFile("TexasTempNone.csv")
temps = lines.map(lambda x: (float(x.split(',')[3]), 1))
tempCounts = temps.reduceByKey(lambda x, y: x + y)

flipped = tempCounts.map(lambda (x,y) : (y,x) )
sortedTemp = flipped.sortByKey()

results = sortedTemp.collect()

for result in results:
    print(result)

Overwriting commonTemp.py


In [2]:
!spark-submit commonTemp.py

17/05/07 23:17:07 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[Stage 0:>                                                          (0 + 2) / 2]/usr/local/Cellar/apache-spark/2.1.0/libexec/python/lib/pyspark.zip/pyspark/shuffle.py:58: UserWarning: Please install psutil to have better support with spilling
/usr/local/Cellar/apache-spark/2.1.0/libexec/python/lib/pyspark.zip/pyspark/shuffle.py:58: UserWarning: Please install psutil to have better support with spilling
(1, 38.5)                                                                       
(1, 38.4)
(1, 38.0)
(1, 46.0)
(1, 60.5)
(1, 60.9)
(1, 62.2)
(1, 41.5)
(1, 70.5)
(1, 66.4)
(1, 73.8)
(1, 61.9)
(1, 62.4)
(1, 83.5)
(1, 56.0)
(1, 74.8)
(1, 85.4)
(1, 77.8)
(1, 69.9)
(1, 52.7)
(1, 39.7)
(1, 40.7)
(1, 54.0)
(1, 86.4)
(1, 61.4)
(1, 74.9)
(1, 38.7)
(1, 51.4)
(1, 43.4)
(1, 78.9)
(1, 54.9)
(1, 57.7)
(1, 53.9)
(1, 37.2)
(1, 63.5)
(1, 54.4)
(1, 42.7)
(1, 43.9)


The most common temperatures are 82.9 and 81. 7, both of them was recorded 12 times from 1895-2017

### QUESTION 4: (Using Hive)
*What is most common HDD and CDD from 1895-2017?

*Under file: "record_1_QUESTION4.txt"

### QUESTION 5: (Using SparkSQL)
What "yearmonth" has a highest PCP, and lowest PCP?

In [3]:
%%file sparkSql.py

from pyspark.sql import SparkSession
from pyspark.sql import Row

import collections

spark = SparkSession.builder.config("spark.sql.warehouse.dir", "file:///C:/temp").appName("SparkSQL").getOrCreate()

def mapper(line):
    fields = line.split(',')
    return Row(sc=int(fields[0]), division=str(fields[1]), yearmonth=int(fields[2]), tavg=float(fields[3]),
               tmin=float(fields[4]), tmax=float(fields[5]), pcp=float(fields[6]), pdsi=float(fields[7]),
                phdi=float(fields[8]),zndx=float(fields[9]), pmdi=float(fields[10]), cdd=float(fields[11]),
              hdd=float(fields[12]), sp01=float(fields[13]), sp02=float(fields[14]), sp03=float(fields[15]),
              sp06=float(fields[16]), sp09=float(fields[17]),sp12=float(fields[18]), sp24=float(fields[19]))

lines = spark.sparkContext.textFile("TexasTempNone.csv")
x = lines.map(mapper)

schemaPcP = spark.createDataFrame(x).cache()
schemaPcP.createOrReplaceTempView("x")

pcps = spark.sql("SELECT yearmonth, pcp FROM x ORDER BY pcp DESC ")

for pcp in pcps.collect():
  print(pcp)


spark.stop()



Overwriting sparkSql.py


In [4]:
!spark-submit sparkSql.py

17/05/07 23:17:25 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Row(yearmonth=201505, pcp=9.05)                                                 
Row(yearmonth=200406, pcp=6.66)
Row(yearmonth=191405, pcp=6.62)
Row(yearmonth=193609, pcp=6.55)
Row(yearmonth=192905, pcp=6.45)
Row(yearmonth=193505, pcp=6.38)
Row(yearmonth=197409, pcp=6.32)
Row(yearmonth=191910, pcp=6.26)
Row(yearmonth=195704, pcp=6.19)
Row(yearmonth=200411, pcp=6.18)
Row(yearmonth=200210, pcp=6.17)
Row(yearmonth=201510, pcp=6.17)
Row(yearmonth=198410, pcp=5.95)
Row(yearmonth=195809, pcp=5.86)
Row(yearmonth=198110, pcp=5.86)
Row(yearmonth=192106, pcp=5.82)
Row(yearmonth=193010, pcp=5.79)
Row(yearmonth=191408, pcp=5.69)
Row(yearmonth=195705, pcp=5.69)
Row(yearmonth=194110, pcp=5.68)
Row(yearmonth=193605, pcp=5.64)
Row(yearmonth=189906, pcp=5.6)
Row(yearmonth=196709, pcp=5.6)
Row(yearmonth=199112, pcp=5.59)
Row(yearmonth=200910, pcp=5.55)
Row(yearmo

Row(yearmonth=200308, pcp=2.09)
Row(yearmonth=189904, pcp=2.08)
Row(yearmonth=191601, pcp=2.08)
Row(yearmonth=193504, pcp=2.08)
Row(yearmonth=197504, pcp=2.08)
Row(yearmonth=198411, pcp=2.08)
Row(yearmonth=191610, pcp=2.07)
Row(yearmonth=191911, pcp=2.07)
Row(yearmonth=193708, pcp=2.07)
Row(yearmonth=195702, pcp=2.07)
Row(yearmonth=196309, pcp=2.07)
Row(yearmonth=198011, pcp=2.07)
Row(yearmonth=199408, pcp=2.07)
Row(yearmonth=200603, pcp=2.07)
Row(yearmonth=189602, pcp=2.06)
Row(yearmonth=190210, pcp=2.06)
Row(yearmonth=190408, pcp=2.06)
Row(yearmonth=191108, pcp=2.06)
Row(yearmonth=198309, pcp=2.06)
Row(yearmonth=189604, pcp=2.05)
Row(yearmonth=190006, pcp=2.05)
Row(yearmonth=197211, pcp=2.05)
Row(yearmonth=197703, pcp=2.05)
Row(yearmonth=189804, pcp=2.04)
Row(yearmonth=190404, pcp=2.04)
Row(yearmonth=191406, pcp=2.04)
Row(yearmonth=193106, pcp=2.04)
Row(yearmonth=193111, pcp=2.04)
Row(yearmonth=198409, pcp=2.04)
Row(yearmonth=199907, pcp=2.04)
Row(yearmonth=190510, pcp=2.03)
Row(year

From the data collected, we found: the highest PCP is 9.05 in 05/2015, and lowest PCP is 0.02 in 10/1952.